### PR TITLE
Manipulability optimization

### PR DESCRIPTION
**Pull request by maltehue**
Added manipulabilityConstraints to be used by the QP Builder and a goal to set these constraints for kinematic chains.
As a result a non zero linear term will be added to the QP that optimizes the manipulability of given expressions.

### CODE REVIEW

File: src/giskardpy/qp/qp_controller.py [hunk 1]
```diff
@@ -224,6 +225,28 @@ def inequality_weight_expressions(self) -> dict:
...
+
+        stacked_exp = cas.vstack([x.expression for x in manip_expressions])
+
+        J = cas.jacobian(stacked_exp, self.get_free_variable_symbols(Derivatives.position))
+
+        grad_traces = []
+        for symbol in self.get_free_variable_symbols(Derivatives.position):
+            JJt = J.dot(J.T)
+            J_dq = cas.total_derivative(J, [symbol], [1])
+            product = cas.Expression(cas.matrix_inverse(JJt)).dot(J_dq).dot(J.T)
```
Comment (ichumuh) [inline src/giskardpy/qp/qp_controller.py:241]: why are you calling cas.Expression? result should already be an expression

File: src/giskardpy/qp/qp_controller.py [hunk 2]
```diff
@@ -224,6 +225,28 @@ def inequality_weight_expressions(self) -> dict:
         error_slack_weights = {f'{c.name}/error': c.normalized_weight() for c in self.inequality_constraints}
         return error_slack_weights
 
+    @profile
+    def linear_weights_expression(self, manip_expressions):
```
Comment (ichumuh) [inline src/giskardpy/qp/qp_controller.py:229]: type hints

File: src/giskardpy/qp/qp_solver.py [hunk 3]
```diff
@@ -204,19 +211,38 @@ def evaluate_functions(self, substitutions: np.ndarray):
...
 
 class QPSWIFTFormatter(QPSolver):
     sparse: bool = True
 
     @profile
     def __init__(self, weights: cas.Expression, g: cas.Expression, lb: cas.Expression, ub: cas.Expression,
                  E: cas.Expression, E_slack: cas.Expression, bE: cas.Expression,
-                 A: cas.Expression, A_slack: cas.Expression, lbA: cas.Expression, ubA: cas.Expression):
+                 A: cas.Expression, A_slack: cas.Expression, lbA: cas.Expression, ubA: cas.Expression,
+                 constraint_jacobian: cas.Expression, grad_traces: [cas.Expression], joints: [cas.Expression]):
```
Comment (ichumuh) [inline src/giskardpy/qp/qp_solver.py:245]: what is the joints parameter used for? can't find relevant code

File: src/giskardpy/qp/qp_solver_qpalm.py [hunk 4]
```diff
@@ -84,6 +87,12 @@ def __init__(self, weights: cas.Expression, g: cas.Expression, lb: cas.Expressio
 
         self.free_symbols_str = [str(x) for x in free_symbols]
 
+        self.init_manipulability_variables(constraint_jacobian, grad_traces)
+        if self.use_manipulability:
```
Comment (ichumuh) [inline src/giskardpy/qp/qp_solver_qpalm.py:91]: why not combine with the other stacked compiled function?

File: src/giskardpy/god_map.py [hunk 5]
```diff
@@ -63,11 +63,14 @@ class GodMap:
...
     neq_constraints: Dict[str, InequalityConstraint]
     derivative_constraints: Dict[str, DerivativeInequalityConstraint]
+    manip_constraints: Dict[str, ManipulabilityConstraint]
     hack: float
     fill_trajectory_velocity_values: bool
     ros_visualizer: ROSMsgVisualization
     free_variables: List[FreeVariable]
     controlled_joints: List[Joint]
+    m_index: List[float]
+    manip_gain: float
```
Comment (ichumuh) [inline src/giskardpy/god_map.py:73]: not sure they should be direct god_map members. Wouldn't they make more sense as qp controller attributes?

File: src/giskardpy/casadi_wrapper.py [hunk 6]
```diff
@@ -2147,3 +2147,11 @@ def substitute(expression, old_symbols, new_symbols):
     new_symbols = Expression([_to_sx(s) for s in new_symbols]).s
     sx = ca.substitute(sx, old_symbols, new_symbols)
     return type(expression)(sx)
+
+
+def matrix_inverse(a):
```
Comment (ichumuh) [inline src/giskardpy/casadi_wrapper.py:2152]: the implementation doesn't match the type hint. you should check if its a transmatrix and then call its inverse method and if it's not, you should turn the result into an Expression

File: src/giskardpy/casadi_wrapper.py [hunk 7]
```diff
@@ -2147,3 +2147,11 @@ def substitute(expression, old_symbols, new_symbols):
...
+    return ca.inv(a.s)
+
+
+def gradient(ex, arg):
```
Comment (ichumuh) [inline src/giskardpy/casadi_wrapper.py:2156]: turn result intro expression

File: src/giskardpy/qp/constraint.py [hunk 8]
```diff
@@ -155,3 +158,14 @@ def is_iterable(self, thing):
     def normalized_weight(self, t):
         weight_normalized = self.quadratic_weight * (1 / self.normalization_factor) ** 2
         return self.horizon_function(weight_normalized, t)
+
+
+class ManipulabilityConstraint:
+    def __init__(self,
+                 name: str,
+                 expression: cas.Expression):
```
Comment (ichumuh) [inline src/giskardpy/qp/constraint.py:166]: are those really all the needed parameters? what about the gain?

File: src/giskardpy/goals/manipulability_goals.py [hunk 9]
```diff
@@ -0,0 +1,61 @@
...
+
+
+class MaxManipulability(Goal):
+    def __init__(self, root_link: str, tip_link: str, gain: float = 0.5,
+                 name: Optional[str] = None,
+                 to_start: Optional[List[Monitor]] = None,
+                 to_hold: Optional[List[Monitor]] = None,
+                 to_end: Optional[List[Monitor]] = None
+                 ):
+        god_map.manip_gain = gain
```
Comment (ichumuh) [inline src/giskardpy/goals/manipulability_goals.py:29]: is that really a global thing?

File: src/giskardpy/goals/manipulability_goals.py [hunk 10]
```diff
@@ -0,0 +1,61 @@
...
+        task.add_manipulability_constraint_vector(names=['x', 'y', 'z'],
+                                                  task_expressions=root_T_tip.to_position()[:3])
+        self.add_task(task)
+        m = symbol_manager.get_symbol(f'god_map.m_index[0]')
+        old_m = symbol_manager.get_symbol(f'god_map.m_index[1]')
+        god_map.debug_expression_manager.add_debug_expression('mIndex_percentualDifference',
+                                                              1 - cas.min(cas.save_division(old_m, m), 1))
+        percentual_diff = 1 - cas.min(cas.save_division(old_m, m), 1)
+        monitor = Monitor(name=f'manipMonitor{tip_link}', crucial=True)
+        monitor.set_expression(cas.less(percentual_diff, 0.01))
```
Comment (ichumuh) [inline src/giskardpy/goals/manipulability_goals.py:47]: 0.01 should be init param?

File: src/giskardpy/goals/manipulability_goals.py [hunk 11]
```diff
@@ -0,0 +1,61 @@
...
+        task.add_to_end_monitor(monitor)
+        self.add_monitor(monitor)
+
+    """
+    This goal maximizes the manipulability of the kinematic chain between root_link and tip_link.
+    This chain should only include rotational joint and no linear joints i.e. torso lift joints or odometry joints.
```
Comment (ichumuh) [inline src/giskardpy/goals/manipulability_goals.py:53]: you can check if there are any non revolute joints in the chain

File: src/giskardpy/goals/goal.py [hunk 12]
```diff
@@ -135,9 +138,13 @@ def get_constraints(self) -> Tuple[Dict[str, EqualityConstraint],
...
                 constraint.name = name
                 self._derivative_constraints[constraint.name] = constraint
+            for constraint in task.get_manipulability_constraint():
+                name = f'{task.name}/{constraint.name}'
+                constraint.name = name
+                self._manip_constraints[constraint.name] = constraint
 
         return self._equality_constraints, self._inequality_constraints, self._derivative_constraints, \
-            self._debug_expressions
+               self._debug_expressions, self._manip_constraints
```
Comment (ichumuh) [inline src/giskardpy/goals/goal.py:147]: return debug expressions last

**Review by ichumuh — CHANGES_REQUESTED** [review]

File: src/giskardpy/goals/manipulability_goals.py (hunk 11 above)
Comment (maltehue) [reply src/giskardpy/goals/manipulability_goals.py:53]: Is there a convenience function do to that?

File: src/giskardpy/qp/qp_solver_qpalm.py (hunk 4 above)
Comment (maltehue) [reply src/giskardpy/qp/qp_solver_qpalm.py:91]: If I do that shouldn't the other two stacked compiled functions be merged together as well? Or with which one should I combine mine?

File: src/giskardpy/qp/qp_solver_qpalm.py [hunk 13]
```diff
@@ -73,13 +75,33 @@ def __init__(self, weights: cas.Expression, g: cas.Expression, lb: cas.Expressio
...
+        len_ub_be_uba_end = len_lb_be_lba_end + ub.shape[0] + bE.shape[0] + ubA.shape[0]
+
+        self.init_manipulability_variables(constraint_jacobian, grad_traces)
+        if self.use_manipulability:
+            self.combined_vector_f = cas.StackedCompiledFunction(expressions=[weights,
+                                                                              lb,
+                                                                              bE,
+                                                                              lbA,
+                                                                              ub,
+                                                                              bE,
```
Comment (ichumuh) [inline src/giskardpy/qp/qp_solver_qpalm.py:90]: you don't need bE twice here. that's probably also just a relict from the past

File: src/giskardpy/qp/qp_solver_qpalm.py [hunk 14]
```diff
@@ -89,9 +111,19 @@ def __init__(self, weights: cas.Expression, g: cas.Expression, lb: cas.Expressio
...
+        if self.use_manipulability:
+            self.calc_det_of_jjt_manipulability(substitutions)
+            self.weights, self.lb, self.bE, self.lbA, self.ub, _, self.ubA, m_grad, m, self.lb_bE_lbA, self.ub_bE_ubA = self.combined_vector_f.fast_call(
+                np.append(substitutions, self.det))
+            self.g = np.zeros(self.weights.shape)
+            self.g[:len(m_grad)] = m_grad
+            god_map.qp_controller.manipulability_indexes[1] = god_map.qp_controller.manipulability_indexes[0]
+            god_map.qp_controller.manipulability_indexes[0] = m
+        else:
+            self.weights, self.lb, self.bE, self.lbA, self.ub, _, self.ubA, self.lb_bE_lbA, self.ub_bE_ubA = self.combined_vector_f.fast_call(
```
Comment (ichumuh) [inline src/giskardpy/qp/qp_solver_qpalm.py:123]: yeah, here the second bE just gets thrown away

**Review by ichumuh — CHANGES_REQUESTED** [review]

File: src/giskardpy/configs/giskard.py [hunk 15]
```diff
@@ -57,7 +57,6 @@ def __init__(self,
             additional_goal_package_paths = set()
         for additional_path in additional_goal_package_paths:
             self.add_goal_package_name(additional_path)
-        god_map.hack = 0
```
[the remark this answers is not in the document]
Comment (maltehue) [reply src/giskardpy/configs/giskard.py]: Sorry, I did not mean to touch the hack :(

File: src/giskardpy/qp/qp_solver_qpalm.py (hunk 14 above)
Comment (maltehue) [reply src/giskardpy/qp/qp_solver_qpalm.py:123]: It seems that the second bE is necessary to resolve self.ub_bE_ubA with a additional view from the StackedCompiledFunction

Comment (ichumuh) [reply src/giskardpy/qp/qp_solver_qpalm.py:123]: true

**Pull request merged**